pypose/module/ipddp.py

```diff
@@ -0,0 +1,573 @@
+        self.y = 0.01 * torch.ones(B + (self.T, ncons), dtype=self.x.dtype, device=self.x.device) # slack variable for infeasible case
+
+        # -------- derivatives --------------------------
+        # terms related with system dynamics
+        self.fx = torch.zeros(B + (self.T, ns, ns), dtype=self.x.dtype, device=self.x.device)
+        self.fu = torch.zeros(B + (self.T, ns, nc), dtype=self.x.dtype, device=self.x.device)
+        self.fxx = torch.zeros(B + (self.T, ns, ns, ns), dtype=self.x.dtype, device=self.x.device)
+        self.fxu = torch.zeros(B + (self.T, ns, ns, nc), dtype=self.x.dtype, device=self.x.device)
+        self.fuu = torch.zeros(B + (self.T, ns, nc, nc), dtype=self.x.dtype, device=self.x.device)
+        # terms related with stage cost
+        self.qx = torch.zeros(B + (self.T, ns), dtype=self.x.dtype, device=self.x.device)
+        self.qu = torch.zeros(B + (self.T, nc), dtype=self.x.dtype, device=self.x.device)
+        self.qxx = torch.zeros(B + (self.T, ns, ns), dtype=self.x.dtype, device=self.x.device)
+        self.qxu = torch.zeros(B + (self.T, ns, nc), dtype=self.x.dtype, device=self.x.device)
+        self.quu = torch.zeros(B + (self.T, nc, nc), dtype=self.x.dtype, device=self.x.device)
+        # terms related with terminal cost
+        self.px = torch.zeros(B + (ns,), dtype=self.x.dtype, device=self.x.device)
+        self.pxx = torch.zeros(B + (ns, ns), dtype=self.x.dtype, device=self.x.device)
+        # terms related with constraint
+        self.cx = torch.zeros(B + (self.T, ncons, ns), dtype=self.x.dtype, device=self.x.device)
+        self.cu = torch.zeros(B + (self.T, ncons, nc), dtype=self.x.dtype, device=self.x.device)
+        # -----------------------------------------------
+
+        self.filter = torch.Tensor([[torch.inf], [0.]])
+        self.err, self.cost, self.logcost = torch.zeros(B, dtype=self.x.dtype, device=self.x.device), torch.zeros(B, dtype=self.x.dtype, device=self.x.device), torch.zeros(B, dtype=self.x.dtype, device=self.x.device)
+        self.step, self.fp_failed, self.stepsize, self.reg_exp_base = 0, False, 1.0, 1.6
+
+        # quantities used in backwardpass
+        self.ky = torch.zeros(B + (self.T, ncons), dtype=self.x.dtype, device=self.x.device)
+        self.Ky = torch.zeros(B + (self.T, ncons, ns), dtype=self.x.dtype, device=self.x.device)
+        self.ks = torch.zeros(B + (self.T, ncons), dtype=self.x.dtype, device=self.x.device)
+        self.Ks = torch.zeros(B + (self.T, ncons, ns), dtype=self.x.dtype, device=self.x.device)
+        self.ku = torch.zeros(B + (self.T, nc), dtype=self.x.dtype, device=self.x.device)
+        self.Ku = torch.zeros(B + (self.T, nc, ns), dtype=self.x.dtype, device=self.x.device)
+        self.opterr, self.reg, self.bp_failed, self.recovery = 0., 0., False, 0
+
+    def getDerivatives(self):
+        self.p_fn.set_refpoint(self.x[...,-1,:], self.u[...,-1,:])
```

Comment:
Same thing here. Can we just directly call the derivative when you are using them? Since you've already defined separate class for the cost and constraints.

Reply:
that will make the backwardpasscompact() overwhelm. Notice that in LQR, the cost-related terms are passed in instead of being computed inside the loop, and there are no second-order dynamics terms and constraint terms, which makes the LQR code much more elegant.